letsencrypt/client/tests/auth_handler_test.py

```diff
@@ -229,6 +235,49 @@ def test_name5_mix(self, mock_chall_path):
         self.assertEqual(
             type(self.handler.client_c["4"][0].chall).__name__, "RecTokenChall")
 
+    @mock.patch("letsencrypt.client.auth_handler.gen_challenge_path")
+    def test_perform_exception_cleanup(self, mock_chall_path):
+        """3 Challenge messages... fail perform... clean up."""
+        # pylint: disable=protected-access
+        self.mock_dv_auth.perform.side_effect = errors.LetsEncryptDvsniError
+
+        challenges = acme_util.get_challenges()
+        combos = acme_util.gen_combos(challenges)
+
+        for i in xrange(3):
+            self.handler.add_chall_msg(
+                str(i),
+                acme_util.get_chall_msg(
+                    str(i), "nonce%d" % i, challenges, combos),
+                "dummy_key")
+
+        path = gen_path(["dvsni", "proofOfPossession"], challenges)
+        mock_chall_path.return_value = path
+
+        # This may change in the future... but for now catch the error
+        self.assertRaises(errors.LetsEncryptAuthHandlerError,
+                          self.handler._satisfy_challenges)
+
+        # Verify cleanup is actually run correctly
+        self.assertEqual(self.mock_dv_auth.cleanup.call_count, 3)
+        self.assertEqual(self.mock_client_auth.cleanup.call_count, 3)
+
+        # Check DV cleanup
+        mock_cleanup_args = self.mock_dv_auth.cleanup.call_args_list
+        for i in xrange(3):
+            # Assert length of arg list was 1
+            arg_chall_list = mock_cleanup_args[i][0][0]
+            self.assertEqual(len(arg_chall_list), 1)
+            self.assertEqual(type(arg_chall_list[0]).__name__, "DvsniChall")
```

Comment:
you should be using `isinstance` here unless you forbid those namedtuple challenge types to be ever subclassed